src/mame/machine/zs01.cpp

```diff
@@ -386,50 +407,94 @@ WRITE_LINE_MEMBER( zs01_device::write_scl )
+							uint16_t msg_crc = ( ( m_write_buffer[ 10 ] << 8 ) | m_write_buffer[ 11 ] );
 
-							if( crc == ( ( m_write_buffer[ 10 ] << 8 ) | m_write_buffer[ 11 ] ) )
+							verboselog( 1, "-> command: %02x (%s)\n", m_write_buffer[ 0 ], ( m_write_buffer[ 0 ] & 1 ) ? "READ" : "WRITE" );
+							verboselog( 1, "-> address: %04x (%02x)\n", data_offset(), m_write_buffer[ 1 ] );
+							verboselog( 1, "-> data: %02x%02x%02x%02x%02x%02x%02x%02x\n",
+								m_write_buffer[ 2 ], m_write_buffer[ 3 ], m_write_buffer[ 4 ], m_write_buffer[ 5 ],
+								m_write_buffer[ 6 ], m_write_buffer[ 7 ], m_write_buffer[ 8 ], m_write_buffer[ 9 ] );
+							verboselog( 1, "-> crc: %04x vs %04x %s\n", crc, msg_crc, crc == msg_crc ? "" : "(BAD)");
+
+							if( crc == msg_crc )
 							{
-								verboselog( 1, "-> command: %02x\n", m_write_buffer[ 0 ] );
-								verboselog( 1, "-> address: %02x\n", m_write_buffer[ 1 ] );
-								verboselog( 1, "-> data: %02x%02x%02x%02x%02x%02x%02x%02x\n",
-									m_write_buffer[ 2 ], m_write_buffer[ 3 ], m_write_buffer[ 4 ], m_write_buffer[ 5 ],
-									m_write_buffer[ 6 ], m_write_buffer[ 7 ], m_write_buffer[ 8 ], m_write_buffer[ 9 ] );
-								verboselog( 1, "-> crc: %02x%02x\n", m_write_buffer[ 10 ], m_write_buffer[ 11 ] );
+								m_configuration_registers[ CONFIG_RC ] = 0; // Reset password fail counter
 
 								switch( m_write_buffer[ 0 ] & 1 )
 								{
 								case COMMAND_WRITE:
-									memcpy( &m_data[ data_offset() ], &m_write_buffer[ 2 ], SIZE_DATA_BUFFER );
-
-									/* todo: find out what should be returned. */
 									memset( &m_read_buffer[ 0 ], 0, sizeof( m_write_buffer ) );
+									m_read_buffer[ 0 ] = STATUS_OK;
+
+									if ( m_write_buffer[ 1 ] == 0xfd )
+									{
+										// Erase
+										std::fill( std::begin( m_data ), std::end( m_data ), 0 );
+										std::fill( std::begin( m_data_key ), std::end( m_data_key ), 0 );
+									}
+									else if ( m_write_buffer[ 1 ] == 0xfe )
+									{
+										// Configuration register
+										memcpy( m_configuration_registers, &m_write_buffer[ 2 ], SIZE_DATA_BUFFER );
+									}
+									else if ( m_write_buffer[ 1 ] == 0xff )
+									{
+										// Set password
+										memcpy( m_data_key, &m_write_buffer[ 2 ], SIZE_DATA_BUFFER );
+									}
+									else if ( data_offset() < sizeof ( m_data ) )
+									{
+										memcpy( &m_data[ data_offset() ], &m_write_buffer[ 2 ], SIZE_DATA_BUFFER );
+									}
+									else
+									{
+										verboselog( 1, "-> unknown write offset: %04x (%02x)\n", data_offset(), m_write_buffer[ 1 ] );
+									}
+
 									break;
 
 								case COMMAND_READ:
-									/* todo: find out what should be returned. */
-									memset( &m_read_buffer[ 0 ], 0, 2 );
+									m_read_buffer[ 0 ] = STATUS_OK;
 
-									switch( m_write_buffer[ 1 ] )
+									if ( m_write_buffer[ 1 ] == 0xfc )
 									{
-									case 0xfd:
+										// TODO: Unknown serial
```

Comment:
I still don't know where this is coming but I wrote up a test on real hardware. The two serial IDs are different. With my Drummania 4th Mix security cart, I get the following values:
0xfc = `91 76 3e 00 01 00 10 6b`
0xfd = `01 a6 9c 94 06 00 00 7b`